nengo_viz/static/viz_slider.js

```diff
@@ -212,12 +226,105 @@ VIZ.Slider.prototype.generate_menu = function() {
     var self = this;
     var items = [];
     items.push(['set range', function() {self.set_range();}]);
+    items.push(['set value', function() {self.user_value();}]);
 
     // add the parent's menu items to this
     // TODO: is this really the best way to call the parent's generate_menu()?
     return $.merge(items, VIZ.Component.prototype.generate_menu.call(this));
 };
 
+VIZ.Slider.prototype.input_set_value = function(ind) {
+    var self = this;
+    if (self.filling_slider_value){
+        return;
+    }
+    else{
+        VIZ.netgraph.menu.hide_any();
+        self.filling_slider_value = true;
+        var text_div = this.sliders[ind].value_display;
+        var original_value = text_div.innerHTML;
+        text_div.innerHTML = '<input id="value_in_field" style=" border:0; outline:0;"></input>';
+        elem = document.getElementById('value_in_field')
+        elem.value = Number(original_value);
+        elem.focus();
+        elem.select();
+        elem.style.width = '3em';
+        elem.style.textAlign = 'center';
+        $(text_div).on('keypress', function (event) {self.submit_value(event.which, ind, text_div);});
+    }
+};
+
+VIZ.Slider.prototype.submit_value = function (button, ind, text_div) {
+    if (button == 13) {
+        var self = this;
+        var slider_range = self.scale.domain();
+        var msg = document.getElementById('value_in_field').value;
+        $(text_div).off('keypress');
+        if (VIZ.is_num(msg)) {
+            self.set_value(ind, VIZ.max_min(Number(msg), slider_range[1], slider_range[0]));
+            return;
+        }
+        else {
+            alert('failed to set value');
+            text_div.innerHTML = 0;
+            self.set_value(ind, VIZ.max_min(0, slider_range[1], slider_range[0]));   
+            return;
+        }
+    }
+};
+
+/** report an event back to the server */
+VIZ.Slider.prototype.notify = function(info) {
+    this.notify_msgs.push(info);
+    
+    // only send one message at a time
+    // TODO: find a better way to figure out when it's safe to send
+    // another message, rather than just waiting 1ms....
+    if (this.notify_msgs.length == 1) {
+        var self = this;
+        window.setTimeout(function() {
+            self.send_notify_msg();
+        }, 1);
+    }
+}
+
+/** send exactly one message back to server
+ *  and schedule the next message to be sent, if any
+ */
+VIZ.Slider.prototype.send_notify_msg = function() {
+    msg = this.notify_msgs[0];
+    this.ws.send(msg);
+    if (this.notify_msgs.length > 1) {
+        var self = this;
+        window.setTimeout(function() {
+            self.send_notify_msg();
+        }, 1);
+    }
+    this.notify_msgs.splice(0, 1);
+}
```

Comment:
Why was this needed?  All this code is mostly copied from VIZ.NetGraph, so if we need this sort of functionality, then we should make it a useful class rather than copying code.